include/ginkgo/core/base/mpi.hpp

```diff
@@ -0,0 +1,1251 @@
+            GKO_ASSERT_NO_MPI_ERRORS(
+                MPI_Win_create(base, size, disp_unit, input_info.get(),
+                               comm->get(), &this->window_));
+        } else if (create_type == win_type::dynamic_create) {
+            GKO_ASSERT_NO_MPI_ERRORS(MPI_Win_create_dynamic(
+                input_info.get(), comm->get(), &this->window_));
+        } else if (create_type == win_type::allocate) {
+            GKO_ASSERT_NO_MPI_ERRORS(
+                MPI_Win_allocate(size, disp_unit, input_info.get(), comm->get(),
+                                 base, &this->window_));
+        } else {
+            GKO_NOT_IMPLEMENTED;
+        }
+    }
+
+    MPI_Win get() { return this->window_; }
+
+    void fence(int assert = 0)
+    {
+        if (&this->window_) {
+            GKO_ASSERT_NO_MPI_ERRORS(MPI_Win_fence(assert, this->window_));
+        }
+    }
+
+    void lock(int rank, int assert = 0, lock_type lock_t = lock_type::shared)
+    {
+        if (lock_t == lock_type::shared) {
+            GKO_ASSERT_NO_MPI_ERRORS(
+                MPI_Win_lock(MPI_LOCK_SHARED, rank, assert, this->window_));
+        } else if (lock_t == lock_type::exclusive) {
+            GKO_ASSERT_NO_MPI_ERRORS(
+                MPI_Win_lock(MPI_LOCK_EXCLUSIVE, rank, assert, this->window_));
+        } else {
+            GKO_NOT_IMPLEMENTED;
+        }
+    }
+
+    void unlock(int rank)
+    {
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Win_unlock(rank, this->window_));
+    }
+
+    void lock_all(int assert = 0)
+    {
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Win_lock_all(assert, this->window_));
+    }
+
+    void unlock_all()
+    {
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Win_unlock_all(this->window_));
+    }
+
+    void flush(int rank)
+    {
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Win_flush(rank, this->window_));
+    }
+
+    void flush_local(int rank)
+    {
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Win_flush_local(rank, this->window_));
+    }
+
+    void flush_all()
+    {
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Win_flush_all(this->window_));
+    }
+
+    void flush_all_local()
+    {
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Win_flush_local_all(this->window_));
+    }
+
+    ~window()
+    {
+        if (this->window_ && this->window_ != MPI_WIN_NULL) {
+            MPI_Win_free(&this->window_);
+        }
+    }
+
+private:
+    MPI_Win window_;
+};
+
+
+/**
+ * Send (Blocking) data from calling process to destination rank.
+ *
+ * @param send_buffer  the buffer to send
+ * @param send_count  the number of elements to send
+ * @param destination_rank  the rank to send the data to
+ * @param send_tag  the tag for the send call
+ * @param comm  the communicator
+ */
+template <typename SendType>
+void send(const SendType* send_buffer, const int send_count,
+          const int destination_rank, const int send_tag,
+          std::shared_ptr<const communicator> comm)
+{
+    auto send_type = detail::get_mpi_type(send_buffer[0]);
```

Comment:
This might be unsafe in debug builds if the data we are trying to send comes from an empty array. A safer alternative IMO would be having a traits struct and using
```suggestion
    auto send_type = detail::mpi_type<SendType>::get_type();
```

Reply:
(talking about dereferencing a nullpointer here)